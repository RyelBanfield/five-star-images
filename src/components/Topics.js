/* eslint-disable no-unused-vars */
import { useEffect, useState } from 'react';
import Select from 'react-select';
import { fetchListOfTopics, fetchPhotosByTopic } from '../API';

function Topics() {
  const [topics, setTopics] = useState(null);
  const [options, setOptions] = useState(null);
  const [selectedTopic, setSelectedTopic] = useState(null);

  useEffect(() => {
    fetchListOfTopics().then((topics) => {
      setTopics(topics);
    });
  }, []);

  useEffect(() => {
    if (topics) {
      setOptions(topics.map((topic) => ({
        value: topic.id,
        label: topic.title,
      })));
    }
  }, [topics]);

  const handleChange = (selection) => {
    fetchPhotosByTopic(selection).then((photos) => {
      setSelectedTopic(photos);
    });
  };

  const clearSelection = (event) => {
    event.preventDefault();
    setSelectedTopic(null);
  };

  return (
    <div className="topics">
      <h1>Topics</h1>
      <Select options={options} placeholder="Select Topic" onChange={(selection) => handleChange(selection.value)} />
      <button type="button" className="clear-button" onClick={(event) => clearSelection(event)}>Clear Search</button>

      {!selectedTopic && (
      <ul>
        {topics && topics.map((topic) => (
          <li key={topic.id}>
            <p>{topic.title}</p>
            <img src={topic.cover_photo.urls.small} alt={topics.alt_description} />
          </li>
        ))}
      </ul>
      )}

      {selectedTopic && selectedTopic.map((photo) => (
        <div key={photo.id}>
          <img src={photo.urls.small} alt={photo.alt_description} />
        </div>
      ))}
    </div>
  );
}

export default Topics;
