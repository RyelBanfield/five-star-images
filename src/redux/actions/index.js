const UPDATE_SEARCH_RESULTS = 'UPDATE_SEARCH_RESULTS';
const UPDATE_RANDOM_PHOTOS = 'UPDATE_RANDOM_PHOTOS';
const UPDATE_TOPICS = 'UPDATE_TOPICS';
const UPDATE_SELECTED_TOPIC_PHOTOS = 'UPDATE_SELECTED_TOPIC_PHOTOS';

export const updateRandomPhotos = (photos) => ({
  type: UPDATE_RANDOM_PHOTOS,
  payload: photos,
});

export const updateHomeSearchResults = (photos) => ({
  type: UPDATE_SEARCH_RESULTS,
  payload: photos,
});

export const updateTopics = (topics) => ({
  type: UPDATE_TOPICS,
  payload: topics,
});

export const updateSelectedTopicPhotos = (photos) => ({
  type: UPDATE_SELECTED_TOPIC_PHOTOS,
  payload: photos,
});
