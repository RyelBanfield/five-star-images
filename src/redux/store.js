import { configureStore } from '@reduxjs/toolkit';
import homeReducer from './slices/homeSlice';
import topicsReducer from './slices/topicsSlice';

const store = configureStore({
  reducer: {
    home: homeReducer,
    topics: topicsReducer,
  },
});

export default store;
