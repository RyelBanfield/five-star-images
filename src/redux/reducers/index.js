import { combineReducers } from 'redux';
import homeReducer from './home';
import topicsReducer from './topics';

const rootReducer = combineReducers({
  home: homeReducer,
  topics: topicsReducer,
});

export default rootReducer;
