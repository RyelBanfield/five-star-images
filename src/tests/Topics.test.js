import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import rootReducer from '../redux/reducers';
import Topics from '../components/Topics';

const store = createStore(rootReducer);

it('renders the main div', () => {
  render(
    <Provider store={store}>
      <Topics />
    </Provider>,
  );
  const main = screen.getByRole('main');
  expect(main).toBeInTheDocument();
});

it('renders heading', () => {
  render(
    <Provider store={store}>
      <Topics />
    </Provider>,
  );
  const heading = screen.getByRole('heading');
  expect(heading).toBeInTheDocument();
});
it('renders correctly', () => {
  const topics = renderer
    .create(
      <Provider store={store}>
        <Topics />
      </Provider>,
    )
    .toJSON();
  expect(topics).toMatchSnapshot();
});
