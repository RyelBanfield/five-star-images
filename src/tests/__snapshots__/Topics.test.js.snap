// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<main
  className="topics"
>
  <h1>
    Topics
  </h1>
</main>
`;
